user-defined-functions/specifications.yaml

```diff
@@ -0,0 +1,200 @@
+# Approach:
+# - Try to express patterns in terms of BNFC (https://bnfc.digitalgrammars.com/) (with slight modifications)
+# - Use <label :: (Type)> expressions to label non-terminals
+
+types:
+  - type: Binding
+    examples:
+      - and ↦ Φ.you
+patterns:
+  - pattern: |
+      MetaBinding. Binding ::= ((<topBindingName :: (MetaId)> "=>")? "\\" <binding :: (MetaId)>)? <path :: (MetaId)> ".." <name :: (MetaId)> "↦" <object :: (Object)>
+    description: |
+      A <path> through formations to <name> where a value after "↦" matches <object>.
+      If <topBindingName> is given, it's equal to the first element of <path>.
+      If <binding> is given, it's equal to the full binding.
+    examples:
+      - pattern: '\?PathHead => \?FullBinding ?Path..?Name ↦ ?Object.world(?Bindings)'
+        context: formation
+        input: "⟦ org ↦ ⟦ eolang ↦ ⟦ main ↦ hello.world(and ↦ Φ.you) ⟧ ⟧ ⟧"
+        matches:
+          - metavar: "?Path"
+            value: ["org", "eolang"]
+          - metavar: "?Name"
+            value: "main"
+          - metavar: "?Object"
+            value: "hello"
+          - metavar: "?Bindings"
+            value: [and ↦ Φ.you]
+          - metavar: "?PathHead"
+            value: ["org"]
+          - metavar: "?FullBinding"
+            value: [org ↦ ⟦ eolang ↦ ⟦ main ↦ hello.world(and ↦ Φ.you) ⟧ ⟧]
+      - pattern: "?Path..?Name ↦ ?DispatchPath..?DispatchName.world"
+        context: application
+        input: "(x ↦ org.eolang.world)"
+        matches:
+          - metavar: "?Path"
+            value: []
+          - metavar: "?Name"
+            value: "x"
+          - metavar: "?DispatchPath"
+            value: ["org"]
+          - metavar: "?DispatchName"
+            value: "eolang"
+  - pattern: |
+      ExpressionObject. Expression ::= Object
+    examples:
+      - ξ
+      - ⟦ a ↦ ?B ⟧
+  - pattern: |
+      ExpressionFunction. Expression ::= Function
+    examples:
+      - $name(?a)
+functions:
+  - pattern: |
+      FunctionMt. Function ::= "$mt" "(" <prefix :: (Expression)> "," <suffix :: (Expression)> ")"
+    description: |
+      Return the string "<prefix>.<suffix>".
+    examples:
+      - applications:
+          - application: $mt(ξ, $mt(ρ, $mt($name(?b_a), write(α0 ↦ ?t))))
+            # Note the result isn't a value because the result contains MetaId-s
+            result: ξ.ρ.?b_a.write(α0 ↦ ?t)
+  - pattern: |
+      FunctionName. Function ::= "$name" "(" <binding :: (MetaId)> ")"
+    description: |
+      Return the name of <binding>.
```

Comment:
Why do we need it if we have `?a` for the name already?

Reply:
We may not have the name in case of `\?A ?a..b -> c`

Reply:
Also, we can use $name to extract all names from a list of bindings.
We may need to do that when we need to check that a name of some binding doesn't belong to that list.